src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -2314,8 +2328,17 @@ def _do_construct_handler(self, node, parent):
         :rtype: :py:class:`psyclone.psyir.nodes.Loop`
 
         :raises NotImplementedError: if the fparser2 tree has an unsupported \
-            structure (e.g. DO WHILE or a DO with no loop control).
-        '''
+            structure (e.g. DO WHILE or a DO with no loop control or a named \
+            DO).
+        '''
+        nonlabel_do = walk(node.content, Fortran2003.Nonlabel_Do_Stmt)[0]
```

Comment:
I could but then I realised that, with the exception of the DO loop, these construct names are just syntactic sugar so we can happily ignore them. The DO loop is different because the construct-name can be referred to by an EXIT or CYCLE statement. Therefore, I've made the do handler a bit more sophisticated so that it now checks for this specific situation.